src/cmap/connect.ts

```diff
@@ -1,28 +1,50 @@
-import net = require('net');
-import tls = require('tls');
-import { Connection } from './connection';
+import * as net from 'net';
+import * as tls from 'tls';
+import { Connection, MongoDBConnectionOptions } from './connection';
 import { MongoError, MongoNetworkError, MongoNetworkTimeoutError } from '../error';
-import { defaultAuthProviders } from './auth/defaultAuthProviders';
+import { AUTH_PROVIDERS } from './auth/defaultAuthProviders';
 import { AuthContext } from './auth/auth_provider';
-import { makeClientMetadata } from '../utils';
+import { makeClientMetadata, ClientMetadataOptions, ClientMetadata } from '../utils';
 import {
   MAX_SUPPORTED_WIRE_VERSION,
   MAX_SUPPORTED_SERVER_VERSION,
   MIN_SUPPORTED_WIRE_VERSION,
   MIN_SUPPORTED_SERVER_VERSION
 } from './wire_protocol/constants';
-const AUTH_PROVIDERS: any = defaultAuthProviders();
-
-function connect(options: any, cancellationToken: any, callback: Function) {
-  if (typeof cancellationToken === 'function') {
-    callback = cancellationToken;
+import type { Callback, CallbackWithType, Document, UniversalError } from '../types';
+import type { EventEmitter } from 'events';
+
+import type { Socket, SocketConnectOpts } from 'net';
+import type { TLSSocket, ConnectionOptions as TLSConnectionOpts } from 'tls';
+
+type UniversalSocket = Socket | TLSSocket;
+
+export function connect(
+  options: MongoDBConnectionOptions,
+  callback: Callback<Connection | UniversalSocket>
```

Comment:
`callback` should only ever return a `Connection`. Can you show me where it returns a socket directly? That looks like a bug!

Reply:
A few lines down, where we call makeConnection, my mistake for not reading the comment "in the error case, `socket` is the originating error event name" but as I look through I think this is outdated, none of the error cases send through a second argument.